presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java

```diff
@@ -92,28 +90,28 @@
     private final SymbolAllocator symbolAllocator = new SymbolAllocator();
     private final Metadata metadata;
     private final SqlParser sqlParser;
-    private final CostCalculator costCalculator;
+    private final Lookup lookup;
 
     public LogicalPlanner(Session session,
             List<PlanOptimizer> planOptimizers,
             PlanNodeIdAllocator idAllocator,
             Metadata metadata,
             SqlParser sqlParser,
-            CostCalculator costCalculator)
+            Lookup lookup)
```

Comment:
I believe this lookup is the StatelessLookup.  This is a weird usage that I agree doesn't fit well with the design, but I think mostly that's because it's not really the right thing to be doing.  This gets passed here because it's needed to create a new Plan, which uses the lookup to create the PlanNodeStatsEstimatesMap. That in turn gets used by the MetricComparator.  But it seems to me like the metric comparator ought to be calculating the stats from the plan itself using its own lookup (especially because it's a testing utility, it's very weird that it introduces something to a Presto class).

The idea behind this design was to leave all recursion and resolution to the lookup and have cost calculator only calculate costs for particular nodes.  You can use a dummy lookup or a real lookup.     I'm not sure what alternative you are suggesting, but for any design you would need lookup somehow in order to get stats/costs because you're recursing through the plan.  I don't think there's a change for stats/cost calculators that would solve this issue.